src/access/pg_tde_tdemap.c

```diff
@@ -330,23 +177,445 @@ tde_sprint_masterkey(const keyData *k)
+	map_entry.relNumber = rlocator->relNumber;
+	map_entry.flags = MAP_ENTRY_VALID;
+	map_entry.key_index = key_index;
+
+	/* Add the entry to the file */
+	if (FileWrite(map_file, &map_entry, MAP_ENTRY_SIZE, 0, WAIT_EVENT_DATA_FILE_WRITE) != MAP_ENTRY_SIZE)
+	{
+		ereport(FATAL,
+				(errcode_for_file_access(),
+					errmsg("Could not write tde map file \"%s\": %m",
+						db_map_path)));
+	}
+
+	/* Let's close the file. */
+	FileClose(map_file);
+
+	return key_index;
+}
+
+int32
+pg_tde_read_map_entry(const RelFileLocator *rlocator, char *db_map_path, const char *master_key_name)
 {
-    char        *rel_file_path;
-    char        *key_file_path = NULL;
-
-    /* We get a relation name for MAIN fork and manually append the
-     * .tde postfix to the file name
-     */
-    rel_file_path = relpathperm(*newrlocator, MAIN_FORKNUM);
-    if (rel_file_path)
-    {
-        key_file_path = psprintf("%s."TDE_FORK_EXT, rel_file_path);
-        pfree(rel_file_path);
-    }
-    return key_file_path;
+	File map_file = -1;
+	int32 key_index = 0;
+	TDEMapEntry map_entry;
+	bool is_new_file;
+	bool found = false;
+
+	/* Open and vaidate file for basic correctness. */
+	map_file = pg_tde_open_file(db_map_path, master_key_name, O_RDONLY, &is_new_file);
+
+	/*
+	 * Read until we find an empty slot. Otherwise, read until end. This seems
+	 * to be less frequent than vacuum. So let's keep this function here rather
+	 * than overloading the vacuum process.
+	 */
+	while(FileRead(map_file, &map_entry, MAP_ENTRY_SIZE, 0, WAIT_EVENT_DATA_FILE_READ))
+	{
+		/* We found a valid entry for the relNumber */
+		if (map_entry.flags == MAP_ENTRY_VALID && map_entry.relNumber == rlocator->relNumber)
+		{
+			found = true;
+			break;
+		}
+
+		/* Increment the key index only if we didn't get a valid slot. */
+		key_index++;
+	}
+
+	/* Let's close the file. */
+	FileClose(map_file);
+
+	/* An entry was expected but none was found */
+	if (found == false)
+	{
+		ereport(FATAL,
+				(errcode(ERRCODE_NO_DATA_FOUND),
+					errmsg("Could not find the required entry for relNumber %d in tde map file \"%s\": %m",
+						rlocator->relNumber,
+						db_map_path)));
+	}
+
+	/* Return the index */
+	return key_index;
+}
+
+/*
+ * Key Data [pg_tde.dat]:
+ * 		header: {Format Version: x}
+ * 		data: {Encrypted Key}
+ * 
+ * Requires a valid index of the key to be written. The function with seek to
+ * the required location in the file. Any holes will be filled when another
+ * job finds an empty index.
+ */
+void
+pg_tde_write_keydata(char *db_keydata_path, const char *master_key_name, int32 key_index, RelKeysData *enc_rel_key_data)
+{
+	File keydata_file = -1;
+	off_t write_pos;
+	bool is_new_file;
+	size_t key_size;
+
+	/* Open and vaidate file for basic correctness. */
+	keydata_file = pg_tde_open_file(db_keydata_path, master_key_name, O_RDWR | O_CREAT, &is_new_file);
+
+	/* Calculate the writing position in the file. */
+	write_pos = TDE_FILE_HEADER_SIZE + (key_index * INTERNAL_KEY_LEN);
```

Comment:
Shouldn't be here 
```suggestion
	write_pos = TDE_FILE_HEADER_SIZE + (key_index * SizeOfRelKeysData(1));
```
As we write whole `InternalKey` (with start_loc, end_loc etc) not just a key